docs/design/storage_classes.md

```diff
@@ -0,0 +1,413 @@
+metadata:
+  name: my-storage-class
+  labels:
+    type: bar
+    color: foo
+description: A very special storage class
+```
+
+#### New API resource or ConfigMap?
+
+During discussion of this topic in
+[kubernetes/17056](https://github.com/kubernetes/kubernetes/pull/17056), it
+was assumed that we would use `ConfigMap` to hold information about storage
+classes.  This would not be impossible to do, but there are several challenges
+due to the fact that `PersistentVolume` is a global resource and `ConfigMap`
+is namespaced.   We would need a virtual resource which appeared to be global
+but which was a view on a namespaced resource -- no such pattern is currently
+implemented anywhere in the system.  For that virtual resource, we would need
+to solve all the permission/policy issues in Kubernetes and downstream
+projects like OpenShift.  We would also likely need to define reserved key
+names for certain attributes of `StorageClass` within the `ConfigMap` `data`
+field.
+
+Even if this approach was considered to be a good design, in order for it to
+be worth it, the time required to implement it would have to be less than the
+time required for a new global resource.  Since this approach is full of
+unknowns, and global resources are well-understood, it makes sense to propose
+that we go the route of the new global resource.
+
+### Labeling volumes with storage class
+
+To indicate that a volume belongs to a storage class, it should be labeled
+with the same labels as the `StorageClass` it is part of.  The following
+volume is part of the `my-storage-class` class:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: ebs-pv-1
+  labels:
+    type: bar
+    color: foo
+spec:
+  capacity:
+    storage: 100Gi
+  accessModes:
+    - ReadWriteMany
+  persistentVolumeReclaimPolicy: Retain
+  awsElasticBlockStore:
+    volumeID: vol-12345
+    fsType: xfs
+```
+
+### Controller Changes
+
+At the time of this writing, the various controllers for persistent volumes
+are in the process of being refactored into a single controller (see
+[kubernetes/24331](https://github.com/kubernetes/kubernetes/pull/24331)).
+
+The resulting controller should be modified to use the new
+`persistentVolumeSelector` field to match a claim to a volume.  In order to
+match to a volume, all criteria must be satisfied; ie, if a label selector is
+specified on a claim, a volume must match both the label selector and any
+specified access modes and resource requirements to be considered a match.
+
+## Examples
+
+Let's take a look at a few examples, revisiting the taxonomy of EBS volumes and regions:
+
+Say that the administrator of a cluster decides to use the `storage-class` and
+`region` labels to model their taxonomy of classes.  The `gold-` classes would look like:
+
+```yaml
+apiVersion: extensions/v1
+kind: StorageClass
+metadata:
+  name: gold-west
+  labels:
+    storage-class: gold
+    region: west
+description: Provisioned IOPS SSD in the west region
```

Comment:
I'm fine with making it an annotation.  For other matters, expressed myself in https://github.com/kubernetes/kubernetes/pull/25413#discussion_r63778413